File: dotnet/CardsProject/Domain/Tournaments/TournamentPrize.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;

namespace CardsProject.Domain.Tournaments;

public enum TournamentPrizePrizeTypeType
{
    Currency,
    Cards,
    BoosterPacks,
    Trophy,
    SeasonPoints,
    Mixed
}

public class TournamentPrize : IValidatableObject
{
    public int Id { get; set; }

    public int PlacementFrom { get; set; } = 0;
    public int PlacementTo { get; set; } = 0;
    public TournamentPrizePrizeTypeType PrizeType { get; set; }
    public decimal Amount { get; set; } = 0.00m;
    public string? Description { get; set; }
    public int? PacksCount { get; set; } = null;
    public int SeasonPoints { get; set; } = 0;

    public int? TournamentId { get; set; }
    [ForeignKey(nameof(TournamentId))]
    public Tournament? Tournament { get; set; }

    // Business operations

    public bool AppliesToPlacement(int placement)
    {
        throw new NotImplementedException("applies_to_placement not implemented");
    }

    public void AwardToPlayer(int playerId)
    {
        throw new NotImplementedException("award_to_player not implemented");
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( PlacementTo >= PlacementFrom ))
            yield return new ValidationResult("placement_to must be greater than or equal to placement_from", new[] { nameof(Id) });
        if (!( PlacementFrom > 0 ))
            yield return new ValidationResult("placement_from must be greater than zero", new[] { nameof(Id) });
        if (!( Amount >= 0m ))
            yield return new ValidationResult("Prize amount must not be negative", new[] { nameof(Id) });
    }
}

File: phoenix/lib/cards_project/tournaments/tournament_prize.ex
defmodule CardsProject.Tournaments.TournamentPrize do
  use Ecto.Schema
  import Ecto.Changeset

  schema "tournament_prizes" do
    field :placement_from, :integer
    field :placement_to, :integer
    field :prize_type, :string
    field :amount, :decimal
    field :description, :string
    field :packs_count, :integer
    field :season_points, :integer, default: 0
    belongs_to :tournament, CardsProject.Tournaments.Tournament

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:placement_from, :placement_to, :amount, :season_points, :prize_type, :description, :packs_count, :tournament_id])
    |> validate_required([:placement_from, :placement_to, :amount, :season_points])
    |> validate_inclusion(:prize_type, ["Currency", "Cards", "BoosterPacks", "Trophy", "SeasonPoints", "Mixed"])
    |> then(fn cs ->
      lv = get_field(cs, :placement_from)
      fv = get_field(cs, :placement_to)
      if not is_nil(lv) and not is_nil(fv) and not (fv >= lv) do
        Ecto.Changeset.add_error(cs, :placement_to, "placement_to must be greater than or equal to placement_from")
      else
        cs
      end
    end)
    |> validate_number(:placement_from, greater_than: 0, message: "placement_from must be greater than zero")
    |> validate_number(:amount, greater_than_or_equal_to: 0, message: "Prize amount must not be negative")
  end

  # ── Business operations ────────────────────────────────────────────

  def applies_to_placement(_record, _placement) do
    # TODO: implement TournamentPrize.applies_to_placement
    {:error, :not_implemented}
  end

  def award_to_player(_record, _player_id) do
    # TODO: implement TournamentPrize.award_to_player
    :ok
  end
end
